src/zc/catalog/index.py

```diff
@@ -44,23 +43,34 @@ def __get__(self, instance, type=None):
         d = instance.__dict__
         if "family" in d:
             return d["family"]
+
         if "btreemodule" in d:
             iftype = d["btreemodule"].split(".")[-1][:2]
             if iftype == "IF":
                 d["family"] = BTrees.family32
             elif iftype == "LF":
                 d["family"] = BTrees.family64
-            else:
+            else: # pragma: no cover
                 raise ValueError("can't determine btree family based on"
                                  " btreemodule of %r" % (iftype,))
         else:
             d["family"] = BTrees.family32
         self._clear_old_cruft(instance)
+
+        if isinstance(instance, persistent.Persistent):
+            # Mutating the dict directly is not guaranteed to
+            # register with the data manager.
+            instance._p_changed = True
```

Comment:
Yuck, write-on-read semantics :(

I've a negative gut reaction to these, but maybe it's harmless in this case?  If multiple threads try to make the same change to the same object's __dict__ from parallel transactions, surely the write conflict can be resolved easily?

(Ideally we'd make sure the family was set in `__init__`, but I suppose all this property magic is for BBB of old objects in existing databases.)